Src: Lesson1; Lesson2; Lesson3; Lesson4;  Test: Lesson2.task1

tyushin opened the pull request.

Commit:
Lesson1
task1

Commit:
lesson2
task1
1 and 2 number

Commit:
Merge remote-tracking branch 'origin/master'

Commit:
lesson2
task1
1 and 2 number

Commit:
Merge remote-tracking branch 'origin/master'

Commit:
src:

lesson1
(all numbers)

lesson2.task1
(numbers:
-"ageDescription"
-"timeForHalfWay"
-"whichRookThreatens"
-"segmentLength")

lesson2.task2
(numbers:
-"isNumberHappy"
-"brickPasses")

lesson3
(number "squareSequenceDigit")

lesson4
(number "mean")

test:

lesson2.task1
(number "segmentLength" add 1 test)

mglukhikh commented:
recheck all

mglukhikh commented:
recheck all

maximpetrov commented on src/lesson1/task1/Simple.kt:
```diff
@@ -102,12 +102,17 @@ fun travelMinutes(hoursDepart: Int, minutesDepart: Int, hoursArrive: Int, minute
  */
-fun accountInThreeYears(initial: Int, percent: Int): Double = TODO()
+fun accountInThreeYears(initial: Int, percent: Int): Double  {
+    val year1: Double =(initial+initial/100.0*percent)
```
Попробуйте упростить это до одного выражения с помощью математических преобразований

maximpetrov replied:
Указывать тип переменной необязательно, если Вы её инициализируете: в этом случае компилятор сам может вывести тип переменной.

maximpetrov commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -33,7 +33,18 @@ fun minBiRoot(a: Double, b: Double, c: Double): Double {
+
+    if ((age > 14) || (age < 11 )) {
+        val ageEnd: Int = (age % 10)
+        if (ageEnd == 1) return ("$age год")
```
Скобки вокруг возвращаемого значения не обязательны

maximpetrov commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -56,7 +80,14 @@ fun timeForHalfWay(t1: Double, v1: Double,
                        rookX1: Int, rookY1: Int,
-                       rookX2: Int, rookY2: Int): Int = TODO()
+                       rookX2: Int, rookY2: Int): Int {
+ if (((kingX == rookX1) && (kingX == rookX2)) || ((kingY == rookY1) && (kingY == rookY2)) || ((kingX == rookX1) && (kingY == rookY2)) || ((kingX == rookX2) && (kingY == rookY1))) return (3)
```
Условия угрозы каждой из ладей королю лучше вынести в отдельные переменные.

maximpetrov commented on src/lesson2/task2/Logical.kt:
```diff
@@ -17,7 +17,14 @@ fun pointInsideCircle(x: Double, y: Double, x0: Double, y0: Double, r: Double) =
+    val number2 = ((number/10) % 10)
+    val number3 = ((number/100) % 10)
+    val number4 = number/1000
+    if ((number1 + number2) == (number3 + number4)) return true
```
Выражение `if (x) return true else return false` эквивалентно `return x`.

maximpetrov commented on src/lesson2/task2/Logical.kt:
```diff
@@ -46,4 +53,7 @@ fun circleInside(x1: Double, y1: Double, r1: Double,
  */
-fun brickPasses(a: Int, b: Int, c: Int, r: Int, s: Int): Boolean = TODO()
+fun brickPasses(a: Int, b: Int, c: Int, r: Int, s: Int): Boolean {
+    if (((a <= r) && ((b <= s)||(c <= s))) || ((b <= r) && ((a <= s)||(c <= s))) || ((c <=r) && ((a <= s) || (b <=s)))) return true
```
Если предварительно вычислить наименьшие стороны кирпича, то условие можно будет записать намного проще.

Commit:
lesson4

Commit:
lesson4

Commit:
Lesson5
1 task

Commit:
Merge remote-tracking branch 'origin/master'

Commit:
Lesson5
1 task

Commit:
Clear

Commit:
Lesson5 (1 task)

Commit:
Lesson5 (1 task)

maximpetrov commented:
Решения новых задач не будут проверяться, пока Вы не исправите старые замечания

Commit:
Переделал lesson1 - lesson2:
lesson1/task1:
-accountInThreeYears (упростил до одного выражения, убрал указания типа переменной)
lesson2/task1:
-ageDescription (убрал скобки вокруг возвращаемого значения)
-whichRookThreatens (условия угрозы каждой из ладей королю вынес в отдельные переменные)
lesson2/task2:
-isNumberHappy (заменил выражение if (x) return true else return false эквивалентным выражением return x)
-brickPasses (придварительно вычислил наименьшие стороны кирпича)

Commit:
Переделал lesson1 - lesson2:
lesson1/task1:
-accountInThreeYears (упростил до одного выражения, убрал указания типа переменной)
lesson2/task1:
-ageDescription (убрал скобки вокруг возвращаемого значения)
-whichRookThreatens (условия угрозы каждой из ладей королю вынес в отдельные переменные)
lesson2/task2:
-isNumberHappy (заменил выражение if (x) return true else return false эквивалентным выражением return x)
-brickPasses (придварительно вычислил наименьшие стороны кирпича)

maximpetrov commented:
recheck all

maximpetrov commented on src/lesson1/task1/Simple.kt:
```diff
@@ -102,12 +102,7 @@ fun travelMinutes(hoursDepart: Int, minutesDepart: Int, hoursArrive: Int, minute
-    val year3: Double =(year2+year2/100.0*percent)
-    return (year3)
-}
+fun accountInThreeYears(initial: Int, percent: Int): Double  = (initial + initial/100.0*percent*3 + initial/10000.0*percent*percent*3 + initial/1000000.0*percent*percent*percent)
```
Это выражение можно упростить гораздо сильнее, пока Вы его скорее усложнили.

maximpetrov commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -34,15 +34,14 @@ fun minBiRoot(a: Double, b: Double, c: Double): Double {
     else {
-        return ("$age лет")
+        val ageEnd: Int = (age % 10)
+        if (ageEnd == 1) return "$age год"
```
Стоит записать цепочку условий с помощью `when` или хотя бы `if ... else if ...`.

maximpetrov commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -81,9 +80,11 @@ fun timeForHalfWay(t1: Double, v1: Double,
-    else if ((kingX == rookX1) || (kingY == rookY1)) return (1)
+    val danger1 = ((kingX == rookX2) || (kingY == rookY2))
+    val danger2 = ((kingX == rookX1) || (kingY == rookY1))
+ if (danger1==true && danger2==true) return (3)
```
- Проверка на  `== true` не имеет смысла.
- Скобки вокруг возвращаемого значения не нужны.

maximpetrov commented on src/lesson4/task1/List.kt:
```diff
@@ -104,22 +104,45 @@ fun buildSumExample(list: List<Int>) = list.joinToString(separator = " + ", post
  */
-fun abs(v: List<Double>): Double = TODO()
+fun abs(v: List<Double>): Double {
+    val count = v.size
```
Зачем заводить эту переменную?

maximpetrov commented on src/lesson4/task1/List.kt:
```diff
@@ -104,22 +104,45 @@ fun buildSumExample(list: List<Int>) = list.joinToString(separator = " + ", post
+    for (i in 0..count-1){
+        result += (v[i] * v[i])
+    }
+    return (Math.sqrt(result))
```
Скобки вокруг возвращаемого значения не нужны

maximpetrov commented on src/lesson4/task1/List.kt:
```diff
@@ -104,22 +104,45 @@ fun buildSumExample(list: List<Int>) = list.joinToString(separator = " + ", post
-fun mean(list: List<Double>): Double = TODO()
+fun mean(list: List<Double>): Double {
+    var result = 0.0
+    var count = 0
```
Зачем Вам считать количество элементов вручную?

maximpetrov commented on src/lesson4/task1/List.kt:
```diff
@@ -104,22 +104,45 @@ fun buildSumExample(list: List<Int>) = list.joinToString(separator = " + ", post
+        count++
+    }
+    if (count == 0) return (0.0)
+    else return (result / count)
```
Скобки вокруг возвращаемого значения не нужны

maximpetrov commented on src/lesson4/task1/List.kt:
```diff
@@ -128,7 +151,17 @@ fun center(list: MutableList<Double>): MutableList<Double> = TODO()
+        }
+        return result
+    }
+    else return (0.0)
```
Скобки вокруг возвращаемого значения не нужны

maximpetrov commented on src/lesson4/task1/List.kt:
```diff
@@ -138,7 +171,18 @@ fun times(a: List<Double>, b: List<Double>): Double = TODO()
+        }
+        return result
+    }
+    else return (0.0)
```
Скобки вокруг возвращаемого значения не нужны

maximpetrov commented on src/lesson4/task1/List.kt:
```diff
@@ -128,7 +151,17 @@ fun center(list: MutableList<Double>): MutableList<Double> = TODO()
-fun times(a: List<Double>, b: List<Double>): Double = TODO()
+fun times(a: List<Double>, b: List<Double>): Double {
+    val count = a.size
+    if (count > 0) {
```
Нужна ли эта проверка? Если элементов 0, то ни одна итерация цикла не выполнится.

maximpetrov commented on src/lesson4/task1/List.kt:
```diff
@@ -138,7 +171,18 @@ fun times(a: List<Double>, b: List<Double>): Double = TODO()
-fun polynom(p: List<Double>, x: Double): Double = TODO()
+fun polynom(p: List<Double>, x: Double): Double {
+    val count = p.size
+    if (count > 0) {
```
Нужна ли эта проверка? Если элементов 0, то ни одна итерация цикла не выполнится.

maximpetrov commented on src/lesson4/task1/List.kt:
```diff
@@ -138,7 +171,18 @@ fun times(a: List<Double>, b: List<Double>): Double = TODO()
+        var result = 0.0
+        for (i in 0..count-1) {
+            val doubleI = i.toDouble()
+            result += (p[i]*Math.pow(x,doubleI))
```
Здесь эффективнее было бы накапливать множитель `x` в степени, а не вычислять его каждый раз заново.

maximpetrov commented on src/lesson4/task1/List.kt:
```diff
@@ -165,7 +208,19 @@ fun factorize(n: Int): List<Int> = TODO()
+    var number = n
+    val result = mutableListOf<Int>()
+    while (number != 1){
+        var i = 2
```
Перебор лучше бы ограничить сверху.

maximpetrov replied:
Ещё в силе

maximpetrov commented on src/lesson4/task1/List.kt:
```diff
@@ -184,7 +265,35 @@ fun convert(n: Int, base: Int): List<Int> = TODO()
-fun convertToString(n: Int, base: Int): String = TODO()
+fun convertToString(n: Int, base: Int): String {
+    val result = mutableListOf<Char>()
+    if (n < base){
```
Зачем здесь этот отдельный случай?

maximpetrov commented on src/lesson4/task1/List.kt:
```diff
@@ -184,7 +265,35 @@ fun convert(n: Int, base: Int): List<Int> = TODO()
+    if (n < base){
+        if (n < 10) {
+            var number = n.toChar()
+            number += 48
```
Вместо "магических чисел" 48 и 87 можно использовать сами символы `'0'` и `'a'`.

maximpetrov replied:
Ещё в силе

maximpetrov commented on src/lesson4/task1/List.kt:
```diff
@@ -174,7 +229,33 @@ fun factorizeToString(n: Int): String = TODO()
  * например: n = 100, base = 4 -> (1, 2, 1, 0) или n = 250, base = 14 -> (1, 3, 12)
  */
-fun convert(n: Int, base: Int): List<Int> = TODO()
+fun convert(n: Int, base: Int): List<Int> {
```
Ваш алгоритм перевода крайне неэффективен. Попробуйте переводить традиционным способом, т.е. начиная с младших цифр.

Commit:
Исправления

Commit:
Исправления

Commit:
Исправления

maximpetrov commented on src/lesson4/task1/List.kt:
```diff
@@ -104,22 +104,45 @@ fun buildSumExample(list: List<Int>) = list.joinToString(separator = " + ", post
  * Если список пуст, не делать ничего. Вернуть изменённый список.
  */
-fun center(list: MutableList<Double>): MutableList<Double> = TODO()
+fun center(list: MutableList<Double>): MutableList<Double> {
```
Ещё в силе

Commit:
Исправления center

tyushin commented:
+fun factorizeToString(n: Int): String {
- val result = factorize(n)
- return result.joinToString(separator = "*")
  +}

2016-11-19 20:14 GMT+03:00 Maxim Petrov notifications@github.com:

> ## _@maximpetrov_ commented on this pull request.
> 
> In src/lesson4/task1/List.kt
> https://github.com/Kotlin-Polytech/KotlinAsFirst2016/pull/131:
> 
> > @@ -165,7 +208,19 @@ fun factorize(n: Int): List<Int> = TODO()
> > - Разложить заданное натуральное число n > 1 на простые множители.
> > - Результат разложения вернуть в виде строки, например 75 -> 3_5_5
> >   */
> >   -fun factorizeToString(n: Int): String = TODO()
> >   +fun factorizeToString(n: Int): String {
> >   -    var number = n
> >   -    val result = mutableListOf<Int>()
> >   -    while (number != 1){
> >   -        var i = 2
> 
> Ещё в силе
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/Kotlin-Polytech/KotlinAsFirst2016/pull/131, or mute
> the thread
> https://github.com/notifications/unsubscribe-auth/AVCZa42bH5mfLVtnFyYwkl8zt3y03pohks5q_y6HgaJpZM4KJ0ga
> .

maximpetrov commented:
Ещё в силе замечание про ограничение перебора сверху

Commit:
Исправления ( Добавлена верхняя граница)

maximpetrov commented on src/lesson4/task1/List.kt:
```diff
@@ -184,7 +244,22 @@ fun convert(n: Int, base: Int): List<Int> = TODO()
+        for (element in resultList) {
+            if (element < 10) {
+                var number = element.toChar()
+                number += 48
```
Всё-таки замените "магические числа" (48 и 97) на соответствующие символы.

maximpetrov commented on src/lesson4/task1/List.kt:
```diff
@@ -157,15 +190,30 @@ fun accumulate(list: MutableList<Double>): MutableList<Double> = TODO()
+    val result = mutableListOf<Int>()
+    while (number != 1){
+        var i = 2
+        while ((number % i != 0) && (i < 2147483648)) {
```
Я имел ввиду не совсем это: нужно ограничить перебор сверху на основании того, какой может быть самый большой простой делитель у конкретного числа (не считая само число)

Commit:
Исправления  (Верхний предел, будет само число, так как оно может быть простым)

maximpetrov commented on src/lesson4/task1/List.kt:
```diff
@@ -157,15 +190,30 @@ fun accumulate(list: MutableList<Double>): MutableList<Double> = TODO()
+    val result = mutableListOf<Int>()
+    while (number != 1){
+        var i = 2
+        while ((number % i != 0) && (i < n)) {
```
Число действительно может быть простым, но этот случай можно рассмотреть отдельно (если не будут найдены другие делители)

maximpetrov commented on src/lesson4/task1/List.kt:
```diff
@@ -184,7 +244,20 @@ fun convert(n: Int, base: Int): List<Int> = TODO()
+                result.add(number + '0'.toInt())
+            } else {
+                val number = element.toChar()
+                result.add(number + 'W'.toInt())
```
Во-первых, `toInt()` не обязательно, сам символ уже является числом (кодом символа). Во-вторых, почему Вы используете `'W'`? используйте более подходящий в данной ситуации символ.

mglukhikh commented:
recheck all

Commit:
Исправления  4 урока
+ 1 сложная задача из 5.

maximpetrov commented on src/lesson4/task1/List.kt:
```diff
@@ -157,15 +190,35 @@ fun accumulate(list: MutableList<Double>): MutableList<Double> = TODO()
+    var number = n
+    val result = mutableListOf<Int>()
+    while (number != 1){
+        var i = 2
```
Не заметил сразу: не стоит на каждой итерации перебирать все множители с начала.

maximpetrov commented on src/lesson4/task1/List.kt:
```diff
@@ -157,15 +190,35 @@ fun accumulate(list: MutableList<Double>): MutableList<Double> = TODO()
+        while (number % i != 0) {
+            i++
+        }
+        if (i <= n/2) {
```
Границу можно сделать меньше

Commit:
Исправлен 4 урок (задание factorize)
+ Сделал немного в 6 уроке.

maximpetrov commented on src/lesson4/task1/List.kt:
```diff
@@ -193,17 +193,17 @@ fun accumulate(list: MutableList<Double>): MutableList<Double> = TODO()
-        if (i <= n/2) {
-        number /= i
-        result.add(i)
+        if (n/i < 2) {
```
От перестановки множителей ничего не меняется. Можно придумать границу меньше, чем `n/2`.

Commit:
Исправления.

maximpetrov commented on src/lesson4/task1/List.kt:
```diff
@@ -195,13 +195,14 @@ fun factorize(n: Int): List<Int> {
-        while (number % i != 0) {
-            i++
-        }
+        // Если в условии ниже поставить "<= 2" то программа будет работать неверно. Я думаю, что "< 2" это крайнее значение
```
Если менять только условие ниже, то программа действительно будет работать неверно, нужно немного поменять алгоритм ещё в одном месте.

Commit:
Задачи 5-6 уроки

maximpetrov commented:
recheck all

Commit:
Исправил 6 урок, сделал 7.

Commit:
Lesson 7
